server/forge/service/service.go

```diff
@@ -0,0 +1,172 @@
+package service
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/jellydator/ttlcache/v3"
+	"github.com/rs/zerolog/log"
+
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge"
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge/bitbucket"
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge/bitbucketdatacenter"
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge/gitea"
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge/github"
+	"go.woodpecker-ci.org/woodpecker/v2/server/forge/gitlab"
+	"go.woodpecker-ci.org/woodpecker/v2/server/model"
+	"go.woodpecker-ci.org/woodpecker/v2/server/store"
+)
+
+type forgeService struct {
+	cache *ttlcache.Cache[int64, forge.Forge]
+	store store.Store
+	ttl   time.Duration
+}
+
+const forgeCacheTTL = 10 * time.Minute
+
+func NewForgeService(_store store.Store) forge.Service {
+	return &forgeService{
+		ttl:   forgeCacheTTL,
+		store: _store,
+		cache: ttlcache.New(ttlcache.WithDisableTouchOnHit[int64, forge.Forge]()),
+	}
+}
+
+func (f *forgeService) getForgeByID(id int64) (forge.Forge, error) {
+	item := f.cache.Get(id)
+	if item != nil && !item.IsExpired() {
+		return item.Value(), nil
+	}
+
+	forgeModel, err := f.store.ForgeGet(id)
+	if err != nil {
+		return nil, err
+	}
+
+	forge, err := setupForge(forgeModel)
+	if err != nil {
+		return nil, err
+	}
+
+	f.cache.Set(id, forge, f.ttl)
+
+	return forge, nil
+}
+
+func (f *forgeService) FromRepo(repo *model.Repo) (forge.Forge, error) {
+	return f.getForgeByID(repo.ForgeID)
+}
+
+func (f *forgeService) FromUser(user *model.User) (forge.Forge, error) {
+	return f.getForgeByID(user.ForgeID)
+}
+
+func (f *forgeService) Main() (forge.Forge, error) {
+	return f.getForgeByID(0) // main forge is always 0 and is configured via environment variables
```

Comment:
I am actually not sure about this one as well. I added it mainly as on login we currently don't know which forge to use. Not sure if we simply want to allow users later on to select one forge from all configured ones on login, but I guess we could simply do so: insert env var configured forge on start, admin / users can login with that one already and later on an admin can add more forges and the users can select their forge on login.

Reply:
Also thought about that. It could be really nice for the UX as we could explain to admins in the UI how to create the oauth app, on the other hand it would require quite some adjustments and we would still want to support setting up a forge via env-variables to support automatic deployments using ansible, ...